gym/vector/async_vector_env.py

```diff
@@ -0,0 +1,482 @@
+
+from gym import logger
+from gym.vector.vector_env import VectorEnv
+from gym.error import (AlreadySteppingError, AlreadyResettingError,
+                       NotSteppingError, NotResettingError, ClosedEnvironmentError)
+from gym.vector.utils import (create_shared_memory, create_empty_array,
+                              write_to_shared_memory, read_from_shared_memory,
+                              concatenate, CloudpickleWrapper, clear_mpi_env_vars)
+
+__all__ = ['AsyncVectorEnv']
+
+
+class AsyncVectorEnv(VectorEnv):
+    """Vectorized environment that runs multiple environments in parallel. It
+    uses `multiprocessing` processes, and pipes for communication.
+
+    Parameters
+    ----------
+    env_fns : iterable of callable
+        Functions that create the environments.
+
+    observation_space : `gym.spaces.Space` instance, optional
+        Observation space of a single environment. If `None`, then the
+        observation space of the first environment is taken.
+
+    action_space : `gym.spaces.Space` instance, optional
+        Action space of a single environment. If `None`, then the action space
+        of the first environment is taken.
+
+    shared_memory : bool (default: `True`)
+        If `True`, then the observations from the worker processes are
+        communicated back through shared variables. This can improve the
+        efficiency if the observations are large (e.g. images).
+
+    copy : bool (default: `True`)
+        If `True`, then the `reset` and `step` methods return a copy of the
+        observations.
+
+    max_retries : int (default: 0)
+        Maximum number of times `AsyncVectorEnv` tries to restart a process if
+        it fails. If `None`, then it always tries to restart a failing process.
+
+    episodic : bool (default: `False`)
+        If `True`, then the environments run for a single episode (until
+        `done=True`), and subsequent calls to `step` have an unexpected
+        behaviour. If `False`, then the environments call `reset` at the end of
+        each episode.
+
+    context : str, optional
+        Context for multiprocessing. If `None`, then the default context is used.
+        Only available in Python 3.
+    """
+    def __init__(self, env_fns, observation_space=None, action_space=None,
+                 shared_memory=True, copy=True, max_retries=0,
+                 episodic=False, context=None):
+        try:
+            self.ctx = mp.get_context(context)
+        except AttributeError:
+            logger.warn('Context switching for `multiprocessing` is not '
+                'available in Python 2. Using the default context.')
+            self.ctx = mp
+        self.env_fns = env_fns
+        self.shared_memory = shared_memory
+        self.copy = copy
+        self.max_retries = max_retries
+        self.episodic = episodic
+        self._num_retries = 0
+
+        if (observation_space is None) or (action_space is None):
+            dummy_env = env_fns[0]()
```

Comment:
Creating the dummy_env can be problematic, it would be better to use the subproc approach where the child is queried for the spaces: https://github.com/openai/baselines/blob/master/baselines/common/vec_env/subproc_vec_env.py#L61

Reply:
The `observation_space` is required to create the shared memory, and the shared memory is required for starting the processes. If we don't create a `dummy_env` in the main process, we would have to start a (dummy) process to get the `observation_space`, and then closing it to create the "real" processes. We wouldn't be able to use that dummy process for further operations (`reset`/`step`/etc...) because it wouldn't have the shared memory.

Reply:
Yeah it looks like multiprocessing can't send shared memory from the child process to the parent :/